Mathlib/Analysis/SumIntegralComparisons.lean

```diff
@@ -160,3 +192,104 @@ theorem MonotoneOn.integral_le_sum_Ico (hab : a ≤ b) (hf : MonotoneOn f (Set.I
     (∫ x in a..b, f x) ≤ ∑ i ∈ Finset.Ico a b, f (i + 1 : ℕ) := by
   rw [← neg_le_neg_iff, ← Finset.sum_neg_distrib, ← intervalIntegral.integral_neg]
   exact hf.neg.sum_le_integral_Ico hab
+
+lemma sum_mul_Ico_le_integral_of_monotone_antitone
+    (hab : a ≤ b) (hf : MonotoneOn f (Icc a b)) (hg : AntitoneOn g (Icc (a - 1) (b - 1)))
+    (fpos : 0 ≤ f a) (gpos : 0 ≤ g (b - 1)) :
+    ∑ i ∈ Finset.Ico a b, f i * g i ≤ ∫ x in a..b, f x * g (x - 1) := by
+  apply sum_Ico_le_integral_of_le (f := fun x ↦ f x * g x) hab
+  · intro i hi x hx
+    simp only [Nat.cast_add, Nat.cast_one, mem_Ico] at hx hi
+    have I0 : (i : ℝ) ≤ b - 1 := by
+      simp only [le_sub_iff_add_le]
+      norm_cast
+      omega
+    have I1 : (i : ℝ) ∈ Icc (a - 1 : ℝ) (b - 1) := by
+      simp only [mem_Icc, tsub_le_iff_right]
+      exact ⟨by norm_cast; omega, I0⟩
+    have I2 : x ∈ Icc (a : ℝ) b := by
+      refine ⟨le_trans (mod_cast hi.1) hx.1, hx.2.le.trans ?_⟩
+      norm_cast
+      omega
+    apply mul_le_mul
+    · apply hf
+      · simp only [mem_Icc, Nat.cast_le]
+        exact ⟨hi.1, hi.2.le⟩
+      · exact I2
+      · exact hx.1
+    · apply hg
+      · simp only [mem_Icc, tsub_le_iff_right, sub_add_cancel]
+        refine ⟨le_trans (mod_cast hi.1) hx.1, hx.2.le.trans ?_⟩
+        norm_cast
+        omega
+      · exact I1
+      · simpa [sub_le_iff_le_add] using hx.2.le
+    · apply gpos.trans
+      apply hg I1 (by simp [hab]) I0
+    · apply fpos.trans
+      apply hf (by simp [hab]) I2
+      exact le_trans (mod_cast hi.1) hx.1
+  · rw [IntegrableOn, ← memℒp_one_iff_integrable]
```

Comment:
This proof part feels quite long. Should we have a version of [MeasureTheory.Memℒp.mul_of_top_left](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Function/LpSeminorm/CompareExp.html#MeasureTheory.Mem%E2%84%92p.mul_of_top_left) for `IntegrableOn`?

Reply:
I don't think it would make a difference: all the steps are pretty specific to the current situation, and the natural API lemmas are already there. The steps are:

* reduction from `Ico` to `Icc`
* argue that it suffices to show that the function is in L^\infty, because that implies L^1
* say that a product of L^\infty functions is in L^\infty
* say that a monotone function on a compact interval is in L^\infty, and that an antitone function on a compact interval is in L^\infty

For each step, we have an API lemma, the proof is a little bit longish because there are several atomic steps (and the proof of antitonicity has a `x - 1` in it so requires shifting the intervals). If you see missing API that would streamline this, I'll be happy to implement it!

Reply:
You can remove step 2 if you add

* product of L^\infty and IntegrableOn is IntegrableOn
* monotone function on a compact interval is (L^p hence) IntegrableOn.

I'll let you decide whether you think this is worth implementing.

Reply:
Good idea: no need to go through L^\infty, just argue that the product of an integrable function and an L^\infty function is integrable. And we already have almost everything for that, I just had to extend a little bit the API. IT's clearly better now, thanks!